### PR TITLE
Improve operation of simple hand launched planes

### PR DESCRIPTION
**Pull request by priseborough**
These changes have been developed following on from code inspection and flying of a strix nano goblin 250gram flying wing with Matek F765-wing controller and M8Q GPS with no airspeed sensor or magnetometer.

1) Airspeed estimate used to scale the pitch and roll loops was too low when launching into a strong wind resulting in limit cycling of the loops until the vehicle turned and the wind speed estimates improved. This PR uses the ARSPD_FBW_MIN and ARSPD_FBW_MAX parameter values to set an airspeed estimate halfway between the two limits and with an uncertainty of half the limit interval. This is used as a 'fake' airspeed measurement to enable faster wind speed estimation as start of flight. Also a zero sideslip assumption is used by default unless inhibited by parameter. These tow measures enable the EKF's to provide a robust wind speed measurement. Only wind speed states are updated by these measurements during this mode of operation.

2) Low ground speed during a hand launch into a strong wind in TAKEOFF mode would result in a nose down pitch angle after launch and flight into ground. When a hand launch is detected by setting of takeoff acceleration of speed threshold limits, a fixed lower pitch angle limit of 5 degrees is used during launch.

3)  Flying without a magnetomer relies on correct operation of the EKF-GSF yaw estimator. A vulnerability in the GSF yaw estimator calculation that could result in underflow of the weighting calculation if large innovations effected all hypothesis filters has been encountered and fixed in the PX4/ecl implementation of this algorithm. This algorithm fix has been incorporated here.

4) When flying without a magnetometer and performing repeat flights on a single power cycle, the EKF yaw drifts on the ground and was not being reset on the following takeoff/launch which can cause loss of navigation and attitude reference. 

**Testing**

Log from flight test at commit https://github.com/ArduPilot/ardupilot/pull/15810/commits/5d47abb0be76f4adfd0707ce039bf3b172f3429a flown on a Strix Nano Goblin with MatekF765-Wing board, GPS and no magnetomer or airspeed sensor.

https://drive.google.com/file/d/15TIeZ61IHtgrbUQXkiPNGzEGRBzMZmFg/view?usp=sharing

### CODE REVIEW

**Comment by priseborough** [comment]
Will rebase and continue testing when #15774 is merged.

**Comment by priseborough** [comment]
During flight in strong winds at or above 0.75 x ARSPD_FBW_MIN , it was found that the vehicle could not be disarmed using the left rudder stick after landing. This was determined to be due to the estimated air speed being equal to wind estimated speed when the vehicle was stationary on ground. The IMU based is_still check has been added. See plot below from todays testing for justification of this change.

Screen Shot 2020-12-13 at 1 20 28 pm

File: ArduPlane/takeoff.cpp
Comment (tridge) [inline ArduPlane/takeoff.cpp:172]: this doesn't look right, comment implies we are applying a min, but we're applying a max

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by Hwurzburg** [comment]
here is the test flight log with current PR...as an added bonus wind was >50% of cruise speed today!....no stability issues...however, got an arm denied before second segment...I had picked it up and was walking to TO position and tried to arm...DCM not aligned...back to bench, let it sit still a couple of seconds and it armed.....third flight I armed while still on bench without issue....also it seems that it takes a long time for EKF3 yaw to align....much longer than I am used to with compass-less planes prevously (probably all EKF2 uses though and before GSF perhaps)....takeoffs are often very curvy before heading locks (see first and second....third was pretty straight)....a lot longer and more radical than before....https://www.dropbox.com/s/cduniqorsk8be2c/1218-ekf3_test.BIN?dl=0
Dropbox

**Comment by Hwurzburg** [comment]
Here is another log from today with really bad hand launch the second time (don't think this will be acceptable)....180 deg climbing circle and a prearm failure on the bench for DCM misalignment again....this also had @rmackay9 's fix to allow a compass less plane to arm without having to manually change EK3_SRC1_YAW to 0 from 1 (default) to allow backwards compatibility with Plane

https://www.dropbox.com/s/6ilo1xu68d9zs8q/1220-ef3_test.BIN?dl=0

**Comment by priseborough** [comment]
@Hwurzburg thank yu for testing this.  I need to understand why we are getting such different results. I will dig into the logs later this week.

**Comment by priseborough** [comment]
@Hwurzburg Can you send me a link to the branch you used to fly this. I used EK3_SRCx_YAW = 0 for my testing.

**Comment by priseborough** [comment]
@Hwurzburg Looking at your log, the underlying issue on the second launch is that the yaw estimator doesn't start running until almost 3 seconds after the throw. 

The reason it worked on your first launch was that the engine run-up whilst armed and on ground resulted in the takeoff_expected flag being set which started the yaw estimator. On the second flight the flag was not set, so the yaw estimator did not start until the EKF's own internal 'is flying' check passed.

The yaw estimator needs to start running shortly before the throw so unless this can be made to happen reliably we may have to revert back to alignment to the GPS velocity vector making the assumption that the vehicle is launched into wind without excessive cross-track drift. With copters we use the arming event because this shortly precedes the launch, however with planes it is possible to arm and have extended period of time on the ground then laps before launch. I will modify the takeoff_expected logic to handle your use case.

File: ArduPlane/takeoff.cpp
Comment (priseborough) [reply ArduPlane/takeoff.cpp]: Fixed

**Comment by priseborough** [comment]
Further testing has exposed a bug in the EKF3 selector algorithm that in the case where the primary EKF is slow to align yaw relative to the secondary EKF, the secondary is not selected.

**Comment by Hwurzburg** [comment]
tested today with single IMU test platform (same as before).....the yaw alignment after a disarm period seems consistent and is a few seconds after launch....tested facing wind and crosswind....no drama...feels like old EKF2 launches
log with replay enabled:  https://www.dropbox.com/s/8ql21awovsbtkvv/1-2-ekf3_test.BIN?dl=0

**Comment by priseborough** [comment]
Loss of GPS dead reckoning with EKF2 enabled is failing the CI test at the moment. Will investigate.

**Comment by priseborough** [comment]
I've pulled out the AP_NavEKF2 changes as all my flight testing has been with EKF3 and the auto-test that is failing is using EKF2. Will include an auto-test for EKF3 dead reckoning.

**Comment by Hwurzburg** [comment]
flew this with 16250 and rebased to master....no issues or bad behavior....

**Comment by priseborough** [comment]
I've flown over 12 hand launches using these changes, the majority after a prior landing and disarm with no unexpected behaviour. I'll send a link to the logs after they they've been uploaded.

File: ArduPlane/servos.cpp
Comment (magicrub) [inline ArduPlane/servos.cpp:560]: This is an unintuitive change. Can you explain it a little more? Also, adding some ( ) would make it more readable there too

Comment (priseborough) [reply ArduPlane/servos.cpp]: Done

**Comment by priseborough** [comment]
Failing CI test passed on my build machine aftrer rebasing on master so have force pushed the rebase

**Comment by magicrub** [comment]
Does this have any impact on heading/mag fusion while an aircraft is temporarily on top of a metal launch vehicle which always messes with heading pre-launch

**Comment by Hwurzburg** [comment]
@tridge @priseborough @peterbarker I think this will pass CI if line 2502 of /ardupilot/libraries/AP_AHRS/AP_AHRS_NavEKF.cpp is deleted...its not needed and CLANG is upset

**Comment by priseborough** [comment]
> Does this have any impact on heading/mag fusion while an aircraft is temporarily on top of a metal launch vehicle which always messes with heading pre-launch

Not that I'm aware of. I've checked that scenario in SITL using SIM_MAG_ALY_Y = 300 and SIM_MAG_ALY_X = -300 to create a 90 deg mag field yaw error on ground and then performing a takeoff using the TAKEOFF mode. This test was run using EKF2 

Screen Shot 2021-01-22 at 1 58 33 pm


and then repeated using EKF3

Screen Shot 2021-01-22 at 1 59 24 pm

Both reset the yaw just after takeoff with normal GPS innovation levels after the reset, the difference being that the EKF3 yaw reset message was different indicating that the yaw reset performed was in response to an error condition.

**Comment by peterbarker** [comment]
@priseborough this is failing some relatively trivial tests.

github rebases PRs on master before running them (on github actions, at least).

So while your recent patch was trivial, and just pushed on top of the existing patches, once it was rebased on top of a whole slew of other changes it's having trouble with home position.

You can replicate it locally by rebasing the branch on master (it rebases cleanly) and running `./Tools/autotest/autotest.py build.Tracker test.Tracker.SetHome`

My *guess* is Randy's change to default us to running EKF3.

**Comment by priseborough** [comment]
@peterbarker I've rebased and made some minor changes to get the SetHome test to pass in my build environment. It was trying to get a home position before the EKF had completed alignment and GPs checks and set its origin.

**Comment by priseborough** [comment]
The cause of the remaining CI failure is eluding me. Running CopterTests1b.ThrottleFailsafe with debugging enabled gave the following back trace output:

Screen Shot 2021-01-23 at 8 14 36 am

It is not clear to me how this is related it the EKF changes

**Comment by peterbarker** [comment]
Right.

So this is catching a bug - kind of.  We're in a situation where home is set but our current location is 0,0,0 and that's a "can't happy".  Well.  Should be.

It bites us with moving from EKF type 10 to EKF2 or EKF3.  If we ever moved from external to EKF2 or EKF3 then it could happen there too, probably.  The nature is relatively simple.  While running under EKF type 10 we always know where we are, so we are perfectly happy to set our home location to our current location, something we do a chunk of on all of the vehicles.  However, if you then switch to EKF type 2 or EKF3 before they're happy with their location, you end in the position that home is set - but the current location comes back as lat=0,lng=0,alt=0, which we consider to be "invalid", so when we go to change its altitude frame (to rel-home or rel-origin) we cark it.  The fix is to reboot after moving back from EKF10 to EKF3.

All that being said - that's probably not what's happening here.  This looks *worse*:


```
AT-0122.9: mav.flightmode=RTL Want=LAND custom=6
APM: EKF3 lane switch 1
APM: EKF primary changed:1
APM: EKF variance
LAND> AT-0123.0: Got mode LAND
AT-0123.1: Waiting for Altitude=5.00 with accuracy 6.00
AT-0123.1: Altitude=61.67 (want 5.000000 +- 6.000000)
.
.
.
AT-0123.9: Altitude=59.05 (want 5.000000 +- 6.000000)
APM: EKF3 IMU1 stopped aiding
APM: EKF3 IMU0 stopped aiding
AT-0124.1: Altitude=57.61 (want 5.000000 +- 6.000000)
.
.
APM: SmartRTL deactivated: bad position
AT-0124.9: Altitude=46.27 (want 5.000000 +- 6.000000)
Mode LAND
AT-0125.1: Altitude=44.66 (want 5.000000 +- 6.000000)
AT-0125.2: Altitude=43.30 (want 5.000000 +- 6.000000)
height 48
AT-0125.3: Altitude=41.69 (want 5.000000 +- 6.000000)
.
.
AT-0127.9: Altitude=11.85 (want 5.000000 +- 6.000000)
AT-0127.9: Attained Altitude=10.960000
AT-0127.9: Waiting for DISARM
AT-0127.9: Waiting for disarm (0.00s so far of allowed 120.00)
AT-0128.2: Waiting for disarm (1.80s so far of allowed 120.00)
AT-0128.3: Waiting for disarm (2.80s so far of allowed 120.00)
AT-0128.6: Waiting for disarm (4.80s so far of allowed 120.00)
AT-0128.8: Waiting for disarm (6.80s so far of allowed 120.00)
AT-0129.0: Waiting for disarm (8.80s so far of allowed 120.00)
AT-0129.3: Waiting for disarm (10.80s so far of allowed 120.00)
AT-0129.6: Waiting for disarm (12.80s so far of allowed 120.00)
AT-0129.8: Waiting for disarm (14.80s so far of allowed 120.00)
height 7
AT-0130.1: Waiting for disarm (16.80s so far of allowed 120.00)
AT-0130.3: Waiting for disarm (18.80s so far of allowed 120.00)
PANIC: Should not be called on invalid location
Running: sh Tools/scripts/dumpstack.sh 3394 >dumpstack_arducopter.3394.out 2>&1
```

Firstly, this test isn't trying to test EKF variances of any sort :-)  So that's bad.

That looks like on disarm we've magically managed to return a location from the EKF of 0,0,0 where we were previously returning a position.  That's bad - I guess we can we treat a core as better than another even if it's never had a location?!

As to "why my PR?" well... if that initial-yaw stuff can change initialisation timings that could explain it...

Needless to say, I don't think this is a false positive, but perhaps "never happened before while theoretically possible" might cover it.

If this test failure is repeatable it is gold.

**Comment by priseborough** [comment]
@peterbarker , thanks for taking a look. I'll dig into it further this week.

**Comment by Hwurzburg** [comment]
@priseborough can I help on this? anxious to see it get in

**Comment by priseborough** [comment]
@Hwurzburg At the moment I'm stuck on the auto test failure. I haven't been able to spend much time on it over the  past two weeks due to other commitments and debugging the issue has proven to be more complex that anticipated due to the interaction between SW elements.

**Comment by priseborough** [comment]
I've rebased to take advantage of CI improvement that save the logs for the failing tests.

**Comment by priseborough** [comment]
The https://github.com/ArduPilot/ardupilot/pull/15810/commits/36a77c4dcbefd2b6c54acf0e91ceaac1921da974 commit is the one causing the copter

         self.start_subtest("Radio failsafe RTL fails into land mode due to bad position.")

subtest to fail

**Comment by priseborough** [comment]
The underlying problem is that in Copter::read_inertia() the position returned by the EKF gets used even if the boolean validity status is false and the lat,lng returned by the EKF is 0,0.

@rmackay9 I don't understand why Copter::read_inertia() needs the vehicle location to update the altitude. Shouldn't the height above origin be sufficient?

**Comment by priseborough** [comment]
@rmackay9 The alternative fix you provided with https://github.com/ArduPilot/ardupilot/pull/15810/commits/724d46f8239eafc12d472d54590ef08b361c7997 solved the problem in a better way so I have pulled it in.

**Comment by priseborough** [comment]
@rmackay9 I've included your patch from https://github.com/ArduPilot/ardupilot/pull/15810/commits/bc3d8059f5df985587815e121eb5db317c2a74de so that the Copter auto tests pass

**Comment by priseborough** [comment]
@tridge the auto test issues have been resolved.

File: libraries/AP_AHRS/AP_AHRS_NavEKF.cpp
Comment (tridge) [inline libraries/AP_AHRS/AP_AHRS_NavEKF.cpp:703]: if ret is false here then we should fallback to the AP_AHRS_DCM::airspeed_estimate() call. Otherwise if EK3 can't provide a velocity but can return a wind then it will return true but not fill in the airspeed

File: libraries/AP_AHRS/AP_AHRS_NavEKF.cpp
Comment (tridge) [inline libraries/AP_AHRS/AP_AHRS_NavEKF.cpp:2552]: this function seems to be completely unused? we don't usually add unused APIs

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
@priseborough I only found minor issues in a review today. I think this should be merged soon?

File: libraries/AP_AHRS/AP_AHRS_NavEKF.cpp
Comment (priseborough) [reply libraries/AP_AHRS/AP_AHRS_NavEKF.cpp:708]: fixed

File: libraries/AP_AHRS/AP_AHRS_NavEKF.cpp
Comment (priseborough) [reply libraries/AP_AHRS/AP_AHRS_NavEKF.cpp]: Was used during initial development. Will remove.

**Comment by priseborough** [comment]
@tridge It includes the patch from https://github.com/ArduPilot/ardupilot/pull/16641 so is blocked until that PR is merged, unless we approve that patch as part of this PR.

**Comment by priseborough** [comment]
This PR is effectively blocked until https://github.com/ArduPilot/ardupilot/pull/16641 is resolved

**Comment by priseborough** [comment]
I've rebased and added a patch so that the getLLH function will never return a zero LLH if the EKF has set an origin.

**Comment by rmackay9** [comment]
@priseborough, great, this should mean that the commit from me, "Copter: current_loc keeps last known location" can be removed and it will still pass the autotests.

**Comment by priseborough** [comment]
@peterbarker The plane frsky telem auto test that failed is passing when I run it on my environment.

File: libraries/AP_AHRS/AP_AHRS_NavEKF.cpp
Comment (tridge) [reply libraries/AP_AHRS/AP_AHRS_NavEKF.cpp:708]: I don't think it is quite fixed. If ret is true and have_inertial_nav() is false then it will still return true but not fill in the airspeed. The fallback at the end for ret==false needs to be outside that if() block

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3.cpp:421]: typo on "disabled"
It would be good to add a sentence to explain why a user might enable this option

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
> @peterbarker The plane frsky telem auto test that failed is passing when I run it on my environment.

it seems to be an intermittent test, for now we just re-run it when it fails

**Comment by peterbarker** [comment]
On Sat, 27 Feb 2021, Andrew Tridgell wrote:

>
>       @peterbarker The plane frsky telem auto test that failed is passing when I run it on my environment.
> 
> it seems to be an intermittent test, for now we just re-run it when it fails

I believe 
https://github.com/ArduPilot/ardupilot/pull/16735#partial-pull-merging

should fix that.

**Comment by tridge** [comment]
> should fix that.

thanks Peter!

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp
Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3.cpp]: I've pushed a couple of commits to address these issues.

**Review by tridge — APPROVED** [review]

**Pull request merged**
